### DISCUSSION TITLE
What is the correct configuration for a 2-way bridge?

### DISCUSSION THREAD

**Discussion by matthewGEA** [emqx/emqx#4534] [Q&A] [answered]
I have failed to establish a 2-way bridge between two EMQx brokers that does not results in an infinite loop of publishes between the two brokers.

The desired configuration is:
- a single broker both publishing and a subscribe to another broker (the bridge is desired to be configured on only one broker, the other just receives the bridge).
- The broker making the bridge will publish its topic without a remote mountpoint, and it will subscribe to the same topic with no local mountpoint.  Both bridge.mqtt.aws.forward_mountpoint and bridge.mqtt.aws.receive_mountpoint will be commented out.  It will be a 1-for-1 sync for a given topic.
- I need the remote broker in the bridge to be aware of when the local broker has published to it so the remote broker does not send back the same data (feedback) as a result of the subscription, causing an infinite loop.
- An example of the publish and subscribe:
    bridge.mqtt.aws.forwards = topic/#
    bridge.mqtt.aws.subscription.1.topic = topic/#
    bridge.mqtt.aws.subscription.1.qos = 0  
- MQTT 3.1.1 is the protocol of the clients, and desired for the bridge.

For this configuration to work, questions:

1. Which versions of EMQx are compatible for this to work?  I tried 4.0.11, 4.2.7, and 4.3-rc.3
2. Can the Rules Engine perform this function, or only the mqtt-bridge plugin?
2. What configuration items do I need to set?
    I have tried multiple combinations of the following parameters without luck:
    - emqx.conf:  mqtt.ignore_loop_deliver
    - plugins/emqx_bridge_mqtt.conf: bridge.mqtt.aws.bridge_mode
3. Is there a way to create a subscription using the rules engine?  I created the publish side of the bridge using the Rules Engine in the GUI, but did not find a way to create a subscription.  This statement in the release notes seems to indicate subscriptions are possible: "The rule engine supports the creation of subscriptions for MQTT bridges" https://github.com/emqx/emqx/releases/tag/v4.2.0
4. Is the publish action in the Rules Engine cluster-aware, meaning that each node in a cluster will publish the data from their clients, so if a node goes down, the bridge can continue?  If this feature is true, it is much more resilient than the bridge plugin, which my understanding is that only a single node in the cluster has the bridge and if this node goes down, the bridge fails.
5. Can you clarify what "bridge.mqtt.aws.bridge_mode = true" actually does with regards to either the MQTT connection or the subscription?

**Comment by matthewGEA** [comment]
To clarify on Question 4 above, I see the "MQTT Subscription" resource, but after successful creation, I only see a client connection on the remote broker, not any subscriptions to the topics that I added to the "MQTT Subscription" resource.

**Comment by zmstone** [comment]
afaik, mqtt bridge has no built in message filter.

I guess you can make the bridge node subscribe to a different topic, then use rule engine to make the forwarding with a source client or user name based filtering.

I am curious about the use case. why is this needed?

**Reply by matthewGEA** [reply]
Mosquitto allows a bridge connection to be specified as IN, OUT, or BOTH. [Refer to this page:](https://mosquitto.org/man/mosquitto-conf-5.html)
`topic pattern [[[ out | in | both ] qos-level] local-prefix remote-prefix]`

I am trying to implement the "BOTH" connection type in EMQx.  So maybe my first post is not describing it correctly, but having a bridge connection do "BOTH" is what I want.  I also want the local and remote prefixes to be "" (empty string; valid configuration for Mosquitto), so the bridge is simply copying the data, not re-mapping.

The use-case for having "BOTH" bridge is so that two clients, one on one broker and the second client on the other broker can subscribe/publish to the same topic and when a publish happens to that topic, both clients can see the data equally.

**Reply by matthewGEA** [reply]
Part of my confusion may be that there is both mqtt.ignore_loop_deliver AND zone.external.ignore_loop_deliver configuration items.  I have only been setting mqtt.ignore_loop_deliver in my tests.  Is this correct?

My understanding is that this parameter allow an EMQx broker to not publish back to the client (via its subscription) the same data it just sent that broker?  From the documentation:
> Whether to ignore the message sent by itself. If it is ignored, it means that EMQ X Broker will not deliver this message to the sender of the message.

It's my understanding that this feature is supposed to help MQTT 3.1.1 to be similar in function to the MQTT 5 called No Local (By Using the non local option the broker will not send you any messages that you have published.)

**Reply by matthewGEA** [reply]
SO, trying "zone.external.ignore_loop_deliver = true" on the remote broker allows the 2-way bridge to work.  The state of mqtt.ignore_loop_deliver on the remote broker doesn't seem to matter, just "zone.external.ignore_loop_deliver = true".

So, can you help me understand the difference between "zone.external.ignore_loop_deliver = true" and "mqtt.ignore_loop_deliver"?

**Reply by zmstone** [reply]
Hi @matthewGEA 

So far I found that no-local is supported by the mqtt client we use 
in the bridge:
https://github.com/emqx/emqtt/blob/master/src/emqtt.erl#L158
However it’s not configurable for the bridge — will follow up with a pr.

WRT `ignore_loop_deliver`, I’m not quite familiar with it, will take a closer look tomorrow, but it worth trying out

**Reply by zmstone** [reply]
Ah sorry you already tried it out and it worked for you. 
great!

**Reply by zmstone** [reply]
Had a quick look at the config.
`mqtt.ignore_loop_deliver` seem to be stale, at least in 4.3
I.e. only per-zone config is in use.

**Reply by matthewGEA** [reply]
Thanks for checking.  So, can you help me understand the difference between "zone.external.ignore_loop_deliver = true" and "mqtt.ignore_loop_deliver"?

Did "mqtt.ignore_loop_deliver" exist prior to zone settings?  Is it supposed to be a global setting for all zones?

**Reply by zmstone** [reply]
Hi @matthewGEA
I believe `mqtt.ignore_loop_deliver` can only be set per-zone.
`mqtt.ignore_loop_deliver` existed before 'zone' was introduced, now it seem to be stale. 
Will send a PR to delete it.

**Comment by zmstone** [comment]
> Which versions of EMQx are compatible for this to work? I tried 4.0.11, 4.2.7, and 4.3-rc.3

4.2 and 4.3

**Comment by zmstone** [comment]
> Can the Rules Engine perform this function, or only the mqtt-bridge plugin?

mqtt bridge can be used as a rule engine action.

With only the bridge, you’ll have to break the loop by using different topics 
Or use two publisher bridges with a filter added (fork the code)

**Comment by zmstone** [comment, marked as answer]
@matthewGEA Thank you for your patience.
I try to sum up the findings in this answer and I'd like to mark it as the answer so future readers do not have to go through the details again.

Scenario:
```
bridging emqx [client] --(publish)-----> target emqx
                   \                           /
                    '-----<--(subscribe)------'
```

emqx's mqtt brdige has the ability to both publish and subscribe.
i.e. The embedded MQTT client can be both a publisher and subscriber.

In order to avoid messages looping back to the bridging node,
`zone.$name.ignore_loop_deliver = true` should be set in the target node.

There is also a bridge config `nl` which is currently missing,
if it is added, it should be able to control the behaviour from the bridging node
when using MQTT v5 (`bridge.mqtt.$name.proto_ver = mqttv5`)

EDIT-1:
`mqtt.ignore_loop_deliver` is used when `zone.$name.ignore_loop_deliver` is missing.

EDIT-2:
To support no-local option from bridging node (not target node),
MQTT v5 no-local config will be added. tracking issue: https://github.com/emqx/emqx/issues/4545

**Reply by matthewGEA** [reply]
This is a great start.  Can we add some additional information to the answer?

1. Version of EMQx where this guidance starts to be applicable?
Through my research, this commit in May 2019 was the start of the existence of "zone.external.ignore_loop_deliver = false"
https://github.com/emqx/emqx/pull/2548/commits/f07d36daa5c01f634091478af4d3b6f071430d7f
I don't know when "mqtt.ignore_loop_deliver" broke, as there is reference to using it from @tigercl Nov 2019: https://github.com/emqx/emqx/issues/3028#issuecomment-553196675

2.  What is the purpose of "bridge.mqtt.$name.bridge_mode"?  I see in EMQx the server receiving the bridge connection correctly shows the BridgeMode value as True or False in the Client details, but I don't know what this actually is supposed to do with regards to the connection or subscription.  Are there differences for MQTT 3.1.1 vs MQTT 5 on the behavior?

3.  Can this same function (2-way bridge) be performed using the rules engine only and creating a Publish Resource and a Subscription Resource?  Would this solution allow the bridge continue in the case of a cluster node failing?

**Reply by matthewGEA** [reply]
Appears that bridge mode has the following effect, the version number of MQTT:
bridge_mode=false:
![image](https://user-images.githubusercontent.com/46966648/114328234-ed1c5500-9b09-11eb-8d28-1cece48b4987.png)
bridge_mode = true:
![image](https://user-images.githubusercontent.com/46966648/114328268-0de4aa80-9b0a-11eb-8461-f564cadd7bd7.png)

If you could help understand whether emqx needs to use version 131 for bridging, or is this just for supporting Mosquitto or other brokers?

**Reply by zmstone** [reply]
I'm afraid I need @tigercl and  @terry-xiaoyu to help answer these questions.

**Reply by zmstone** [reply]
`mqtt.ignore_loop_deliver` is used when `zone.$name.ignore_loop_deliver` is missing. (added to the answer)

**Reply by tigercl** [reply]
Because our `bridge_mode` field is placed in the upper four bits of the protocol version field of the CONNECT packet, although the protocol specification is not followed, this is also a common practice of MQTT Broker. So this also leads to the protocol version displayed in Wireshark is 131 when `bridge_mode` is configured as true.

**Reply by matthewGEA** [reply]
Thank you for clarifying that "bridge_mode" sets the MSB bit of the version.  Can you clarify if there is any function for this when EMQx is connecting by bridge to another EMQx broker?  Any guidance on what function it performs will be useful.

**Reply by tigercl** [reply]
It will determine whether the Broker retains the value of the retain field when forwarding publish packets to the bridged Broker.

**Reply by nosovk** [reply]
It seems that mqtt.ignore_loop_deliver stopped working at latest emqx release. I see only one solution for now - is to separate in and out to different topics :(

**Reply by adampassey** [reply]
We opted to solve this using an egress bridge and a couple of rules until `NO LOCAL` is supported:

**Bridge**
The egress bridge forwards all messages from the `outgoing` topic to the `incoming` topic on the bridge. Messages have been decorated to include their original topic and payload (the `publish_messages_to_bridge` rule handles this).

**Replicate Bridged Messages Internally Rule** 
`replicate_bridged_messages_internally`
Then we have a rule that forwards all messages received on the `incoming` topic to the original topic, by extracting it from the decorated message. It will then republish the message to the extracted topic and the original payload.
```
SELECT payload.payload as p, payload.topic as t FROM "incoming"
```
It then publishes the payload `${p}` to the original topic `${t}`.

**Publish Messages to Bridge Rule** 
`publish_messages_to_bridge`
This rule is responsible for forwarding messages published to any topic to the bridge. It decorates the message so the receiver can extract the original topic and payload. It also checks the `clientid` and `topic` of the message to ensure it doesn't get in an infinite loop:
```
SELECT * FROM "#" 
    WHERE 
        topic != 'outgoing'
        AND regex_match(clientid, 'replicate_bridged_messages_internally') = false
        AND regex_match(clientid, '^bridge') = false
```
It's important to check the `clientid` in this rule as the `replicate_bridged_messages_internally` message becomes the `clientid` when the rule replicates messages internally. Checking against `^bridge` should be replaced by the name of your bridge. 

This rule then decorates the message when published:
```json
{
  "topic": "${topic}",
  "payload": "${payload}"
}
```
---
With this setup it's important to restrict others from publishing to `incoming` and `outgoing` topics, via ACL.

**Reply by matthewGEA** [reply]
What a solution!

I wish that Emqx could natively support "aggregation links" as a type of bridge that would allow bi-directional traffic between clusters without loops.

**Comment by matthewGEA** [comment]
I think I have found an additional bug.  Please verify and I'll open a separate issue:

The following is true with all brokers are 4.2.9 -OR- all brokers are 4.3-rc.3

Given there are two separate "local" emqx brokers connecting with the bridge-plugin to a single remote (aggregating) broker
Given the topic is the same for the forward and subscribe on both local brokers with the configured bridges
Given that the remote (aggregating) broker has the following configured: "zone.external.ignore_loop_deliver = true"
When I publish a topic within the bridged scope to any broker, there is an infinite loop that occurs.
NOTE: the infinite loop starts no matter which broker I publish to: the remote (aggregating) broker, or either of the two local brokers (has bridges configured)

Given the infinite loop condition exists
When I stop the bridge plugin on one of the two brokers, the infinite loop stops.
I can then continue to publish to the topic again and no infinite loop occurs.
NOTE: the infinite loop stops no matter which broker I select to disable the bridge plugin

Given the topic is the same for forward and subscribe on one local broker, but is a different topic for forward and subscribe on the other broker
When I publish a topic within the bridged scope (either topics) then an infinite loop DOES NOT occur.

**Reply by tigercl** [reply]
Can you provide your config in `emqx_bridge_mqtt.conf`?

**Reply by matthewGEA** [reply]
Below are the relevant settings.  All others are default per version 4.3-rc.3:

```
bridge.mqtt.aws.address = <ip address>:1883
bridge.mqtt.aws.proto_ver = mqttv3
bridge.mqtt.aws.start_type = auto
#bridge.mqtt.aws.bridge_mode = false
bridge.mqtt.aws.clientid = bridge_aws
bridge.mqtt.aws.clean_start = true
bridge.mqtt.aws.username = admin
bridge.mqtt.aws.password = public
bridge.mqtt.aws.forwards = test/#
#bridge.mqtt.aws.forward_mountpoint = bridge/aws/${node}/
bridge.mqtt.aws.subscription.1.topic = test/#
bridge.mqtt.aws.subscription.1.qos = 0
## bridge.mqtt.aws.receive_mountpoint = receive/aws/
```

**Reply by matthewGEA** [reply]
Using these settings on both local brokers results in the infinite loop.

Using the settings above on one broker and changing the 2nd broker to use `forwards` and `subscription.1.topic` with `test1/#` works as normal.

**Reply by matthewGEA** [reply]
@tigercl Is there any additional information you need?  Have you been able to reproduce?

**Reply by matthewGEA** [reply]
Opened #4619 to discuss this issue.
